crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -409,18 +337,51 @@ impl CheckpointExecutorEventLoop {
                                 i
                             )
                         });
-                    self.schedule_checkpoint(checkpoint, pending)?;
+
+                    if !self.should_schedule_checkpoint(&checkpoint, epoch_store.epoch(), pending) {
+                        break;
+                    }
+
+                    self.schedule_checkpoint(checkpoint, pending, epoch_store.clone())?;
                 }
             }
         }
 
         Ok(())
     }
 
+    fn should_schedule_checkpoint(
+        &self,
+        checkpoint: &VerifiedCheckpoint,
+        epoch: EpochId,
+        pending: &CheckpointExecutionBuffer,
+    ) -> bool {
+        let task_limit = self.config.checkpoint_execution_max_concurrency;
+        let checkpoint_epoch = checkpoint.epoch();
+
+        if pending.len() >= task_limit {
+            return false;
+        }
+
+        match checkpoint.epoch() - epoch {
+            0 => true,
+            1 => false,
+            _ => {
+                panic!(
+                    "Checkpoint epoch {:?} of checkpoint {:?} does not match current epoch {:?}",
+                    checkpoint_epoch,
+                    checkpoint.sequence_number(),
+                    epoch,
+                );
+            }
+        }
+    }
+
```

Comment:
I don't quite get this, why not just `checkpoint.epoch() == epoch`?

Reply:
Oh actually this is a bug. I was assuming that either checkpoint epoch is the same as the current epoch or current epoch + 1 (all other values being errors). This is not true. We could be leading in state sync by multiple epochs.